### PR TITLE
html sanitized team name is stored in database #2801

### LINKED ISSUE #2801: html sanitized team name is stored in database
html sanitization should only be done when displaying as html.
Need to find out what other fields are handled this way as well.

### PR DESCRIPTION
**Pull request by bamboo3250**
fix #2801

I removed sanitization from StudentAttributes class so there is a little change in the order of team names in the test output. 
Before, it was:
Team &amp;gt;&amp;&#35;39;&amp;quot;&amp;lt; 1
Team 2
Team 3

Then it became:
Team 2
Team 3
Team >'"< 1

### CODE REVIEW

**Comment by kanghj** [comment]
I think in some of the earlier issues, such as #2616, we added code to un-sanitize the student's team in our logic, could you make a new issue to remove them now that they are not needed?

File: src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import teammates.common.util.StringHelper;
+import teammates.common.util.Utils;
+import teammates.logic.core.StudentsLogic;
+import teammates.storage.datastore.Datastore;
+
+public class DataMigrationForSanitizedData extends RemoteApiClient {
+
+    public static void main(String[] args) throws IOException {
+        DataMigrationForSanitizedData migrator = new DataMigrationForSanitizedData();
+        migrator.doOperationRemotely();
```
Comment (kanghj) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java]: Could you add a boolean variable for previewing the modifications of the script? So that before running the data migration, we can do a verification of the changes.

File: src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        return s;
+    }
+    
+    private void fixSanitizedDataForStudent(StudentAttributes student) {
+        student.comments = fixSanitization(student.comments);
+        student.course = fixSanitization(student.course);
+        student.email = fixSanitization(student.email);
+        student.googleId = fixSanitization(student.googleId);
+        student.key = fixSanitization(student.key);
```
Comment (kanghj) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java]: key shouldn't need to be fixed

File: src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        student.lastName = fixSanitization(student.lastName);
+        student.name = fixSanitization(student.name);
+        student.section = fixSanitization(student.section);
+        student.team = fixSanitization(student.team);
+        
+        try {
+            StudentsLogic.inst().updateStudentCascade(student.email, student);
```
Comment (kanghj) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java]: This will be a long running script. I think you should only write to the database if there are changes.

Comment (kanghj) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java]: It doesn't need to cascade right? I think only the sanitized data is only on StudentAttributes?

File: src/main/java/teammates/ui/controller/FeedbackSubmissionEditPageData.java [hunk 4]
```diff
@@ -170,10 +170,10 @@ public void setRegisterMessage(String registerMessage) {
         
         for (Map.Entry<String, String> pair : emailNamePair.entrySet()) {
             result.add(
-                "<option value=\"" + pair.getKey() + "\"" 
+                "<option value=\"" + sanitizeForHtml(pair.getKey()) + "\""
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/FeedbackSubmissionEditPageData.java]: Why is this needed?

Comment (kanghj) [inline src/main/java/teammates/ui/controller/FeedbackSubmissionEditPageData.java]: Oh, I understand. Ignore this then

File: src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java (hunk 3 above)
Comment (bamboo3250) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java]: I only found `updateStudentCascade`. Is there any other way to update the student?

Comment (kanghj) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java]: If there isn't one, you can write one :D. Should be identical to updateStudentCascade, but without the modifications to the other entity types

Comment (bamboo3250) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedData.java]: @kanghj I just added updateStudent function. Please review it :D

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 5]
```diff
@@ -207,6 +207,24 @@ public boolean isStudentsInSameTeam(String courseId, String student1Email, Strin
         return isStudentInTeam(courseId, student1.team, student2Email);
     }
     
+    /*
+     * update students without affecting other classes. Use with caution!!!
+     */
+    public void updateStudent(String originalEmail, StudentAttributes student) throws InvalidParametersException,
```
Comment (kanghj) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Can we move this into DataMigrationForSanitizedDataInStudentAttributes? This method should never be used again so it will be better if it's not part of the StudentsLogic class to prevent future developers from getting confused.

**Comment by kanghj** [comment]
Another thing to check: in <code>StudentAttributes#sanitizeForSaving</code> lines 344 to 350, should these lines also be removed?

**Comment by bamboo3250** [comment]
Yes, I did remove them, didn't I? We shouldn't sanitize anything when storing into database.

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 5 above)
Comment (bamboo3250) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Okay, changes made. Could you review again? :smile:

**Comment by kanghj** [comment]
Oops sorry, I was thinking of AccountsAttributes and not StudentAttributes. In AccountsAttributes' <code>sanitizeForSaving</code>, the googleId seems to be sanitized too? Although it should be quite safe to assume that googleIds wouldn't be affected by the sanitization issue (since most email providers don't allow html characters in emails), would it better if we remove the sanitization on the googleId in both AccountAttributes and StudentAttributes, which is used for joining students and accounts?

**Comment by bamboo3250** [comment]
Okay, i'll do for both AccountAttributes and StudentAttributes. Ideally, we should publish new version then immediately run the script (But we still can preview sanitized content first).

**Comment by bamboo3250** [comment]
@damithc Could you run the script first? It will show all sanitized data that will be clean after we change `isPreview` to `false`. As discussed with @kanghj , we want to make sure the email does not contain any html character such as ampersand and single quote. Otherwise, it will affect AccountAttributes and InstructorAttributes as well since those classes connected by user's email address. If so, further amendment should be done to make sure they will work well together.

**Comment by bamboo3250** [comment]
@damithc I just updated the script to help you check the number of emails and google ids containing HTML characters. After running the script, it will show a report at the end of the output. As long as those two numbers are all 0, we can proceed further without problem.

screenshot 2016-01-18 03 20 59

**Comment by damithc** [comment]
Does it also show which emails/Google IDs needed sanitizing? The script will take very long to run as we have 70K users, so it should show all possible details after one run.

**Comment by bamboo3250** [comment]
Yes it does. It will show all data of StudentAttributes that are sanitized and stored in the database. The image above show that those students have their `comments` sanitized. Some others have their team names sanitized as well. The script will display all of them then sum up in a short report at the bottom.

File: src/main/java/teammates/client/scripts/DataMigrationForSanitizedDataInStudentAttributes.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                System.out.println("new team: " + fixSanitization(student.team));
+            }
+            System.out.println();
+        }
+        return hasSanitizedData;
+    }
+
+    private boolean isSanitizedString(String s){
+        
+        if ((s.indexOf('<') >= 0) || (s.indexOf('>') >= 0) || (s.indexOf('\"') >= 0) ||
```
Comment (damithc) [inline src/main/java/teammates/client/scripts/DataMigrationForSanitizedDataInStudentAttributes.java]: In the live server, there are cases where `s` is `null`.

**Comment by damithc** [comment]
Here is the result for the live server
```
There are/is 0 sanitized email(s)!
There are/is 0 sanitized Google Id(s)!
There are/is 13191 student(s) with sanitized data!
```

**Comment by bamboo3250** [comment]
Thanks. I will fix the above bug. 
@kanghj I guess there will be no problem to do StudentAttributes, InstructorAttributes and AccountAttributes separately?

**Comment by kanghj** [comment]
Yes, it's fine to do them separately then. Remember to make issues for them.

**Comment by bamboo3250** [comment]
@kanghj ok, this PR is ready to be reviewed again

**Comment by kanghj** [comment]
Looks good to me, @bamboo3250. Sorry that I missed this out.

**Comment by bamboo3250** [comment]
I recommend we run the script after deploying but the closer they are, the better to avoid showing sanitized-twice strings.

**Comment by kanghj** [comment]
I don't think the strings will be sanitized twice, <code>sanitizeForHtml</code> should be smart enough to handle that? Remember to fix your merge conflicts. If there are non-trivial conflicts, do get them reviewed too.

**Comment by bamboo3250** [comment]
Now we are using <c:out> to do sanitization in the frontend instead of sanitizeForHtml in the backend so the strings will still be sanitized again by <c:out> in JSTL.

**Comment by kanghj** [comment]
Oh, you are right. I forgot that was changed.
